ethers-middleware/src/policy.rs

```diff
@@ -17,34 +19,6 @@ pub trait Policy: Sync + Send + Debug {
     async fn ensure_can_send(&self, tx: TypedTransaction) -> Result<TypedTransaction, Self::Error>;
 }
 
-/// A policy that does not restrict anything.
-#[derive(Debug, Clone, Copy)]
-pub struct AllowEverything;
-
-#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
-#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
-impl Policy for AllowEverything {
-    type Error = ();
-
-    async fn ensure_can_send(&self, tx: TypedTransaction) -> Result<TypedTransaction, Self::Error> {
-        Ok(tx)
-    }
-}
-
-/// A policy that rejects all transactions.
-#[derive(Debug, Clone, Copy)]
-pub struct RejectEverything;
-
-#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
-#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
-impl Policy for RejectEverything {
-    type Error = ();
-
-    async fn ensure_can_send(&self, _: TypedTransaction) -> Result<TypedTransaction, Self::Error> {
-        Err(())
-    }
-}
-
 /// Middleware used to enforce certain policies for transactions.
 #[derive(Clone, Debug)]
 pub struct PolicyMiddleware<M, P> {
```

Comment:
need a policy combinator, otherwise all policies will be evaluated sequentially

Reply:
```
type TypelessError = Box<dyn Error + Send + Sync>;
type TypelessPolicy = Box<dyn Policy<Error = TypelessError>>;

#[derive(Debug)]
struct PolicyCombinator {
    inner: Vec<TypelessPolicy>,
}

#[async_trait]
impl Policy for PolicyCombinator {
    type Error = TypelessError;

    async fn ensure_can_send(&self, tx: TypedTransaction) -> Result<TypedTransaction, Self::Error> {
        let futs = self.inner.iter().map(|i| i.ensure_can_send(tx.clone()));

        // assumes that no policy modifies the tx request
        Ok(join_all(futs)
            .await
            .into_iter()
            .collect::<Result<Vec<_>, _>>()?
            .into_iter()
            .nth(0)
            .unwrap())
    }
}

```

Reply:
question:

is there a way to structure the `Policy` trait that does not require type-erasing the error in order to build a combinator?

Reply:
made it easier to type erase things in the meantime

```

/// A buildable policy combinator
#[derive(Debug)]
pub struct PolicyCombinator {
    inner: Vec<TypelessPolicy>,
}

#[async_trait]
impl Policy for PolicyCombinator {
    type Error = TypelessError;

    async fn ensure_can_send(&self, tx: TypedTransaction) -> Result<TypedTransaction, Self::Error> {
        let futs = self.inner.iter().map(|i| i.ensure_can_send(tx.clone()));

        // assumes that no policy modifies the tx request
        Ok(join_all(futs)
            .await
            .into_iter()
            .collect::<Result<Vec<_>, _>>()?
            .into_iter()
            .nth(0)
            .unwrap())
    }
}

impl PolicyCombinator {
    /// Add a policy to this combinator
    pub fn with<P, E>(mut self, policy: P) -> Self
    where
        P: Policy<Error = E> + 'static,
        E: std::error::Error + Send + Sync + 'static,
    {
        let policy = Erased { inner: policy };

        self.inner.push(Box::new(policy));
        self
    }

    /// Converts into a PolicyMiddleware
    pub fn wrap<M>(self, middleware: M) -> PolicyMiddleware<M, PolicyCombinator>
    where
        M: Middleware,
    {
        PolicyMiddleware::new(middleware, self)
    }
}

#[derive(Debug)]
struct Erased<P, E>
where
    P: Policy<Error = E>,
    E: std::error::Error + Send + Sync + 'static,
{
    inner: P,
}

#[async_trait]
impl<P, E> Policy for Erased<P, E>
where
    P: Policy<Error = E>,
    E: std::error::Error + Send + Sync + 'static,
{
    type Error = TypelessError;

    async fn ensure_can_send(&self, tx: TypedTransaction) -> Result<TypedTransaction, Self::Error> {
        Ok(self.inner.ensure_can_send(tx).await?)
    }
}
```

Reply:
Cool, 
just to make sure I understand that right, this is aimed to make it easier to combine different `Policy` implementations, such as `MemoryPolicy` which is treated as a combination of several `Rules` which represent cases you might want to check against. Since they don't require async, I'd stick to the current implementation?

The PolicyCombinator looks like a great way to then combine custom Policy impls, for example combine multiple database lookups?

looking at this
```rust 
        let futs = self.inner.iter().map(|i| i.ensure_can_send(tx.clone()));

        // assumes that no policy modifies the tx request
        Ok(join_all(futs)
            .await
            .into_iter()
            .collect::<Result<Vec<_>, _>>()?
            .into_iter()
            .nth(0)
            .unwrap_or(tx))

```
I'm thinking the `ensure_can_send` needs some work, maybe we make this take an `Arc<Tx>` to ensure it wont get modified and to remove the need for cloning the tx for each Policy

The policy middleware will then be something like:
```
tx = Arc::new(tx);
select first failing policies(tx.clone());
inner(tx.unwrap_or_clone)
```

Reply:
what about a COW?

Reply:
also heres a version that returns all errors instead of only 1

```
        let (oks, errs): (_, Vec<_>) = join_all(futs).await.into_iter().partition(|x| x.is_ok());

        if errs.is_empty() {
            Ok(oks.into_iter().nth(0).unwrap_or(Ok(tx)).unwrap())
        } else {
            Err(errs.into_iter().map(|r| r.unwrap_err()).collect())
        }
        ```

Reply:
You can also `collect::<Result<_, _>, _>?` and this will return on the first error occurred.

Reply:
oh this is in case you specifically want to return ALL errors in a `Result<_, Vec<Error>>`